### ISSUE TITLE
BayazitDecomposer: IndexOutOfBoundsException for a triangle using PolygonShape

### ISSUE THREAD

**Issue by rjschof** [web-cat/sofia-graphics#3]
I was working on a new design for the TriangleShape class, as you suggested. I ran into the following problem. Also, I had another idea for the design of the class, which I will explain on the page for the pull request shortly. 

I used the following code to try and create an instance of a PolygonShape object on the screen for an app: 

``` java
PolygonShape pShape = new PolygonShape(50, 50, 25, 100, 75, 100);
pShape.setFillColor(Color.maroon);
pShape.setColor(Color.black);
add(pShape);
```

The following stack trace was printed as a result of executing the above code. I spent some time attempting to diagnose the problem, but I ultimately do not know enough about how this class is supposed to work to come up with a solution myself. I feel like the problem occurs because the class that calls it (Polygon.java) is attempting to decompose something that is already in its simplest form. That's just my guess, though.

```
 FATAL EXCEPTION: main
 Process: org.awesomeco.trianglesmash, PID: 2581
 java.lang.IndexOutOfBoundsException: Invalid index 1, size is 1
     at java.util.ArrayList.throwIndexOutOfBoundsException(ArrayList.java:255)
     at java.util.ArrayList.get(ArrayList.java:308)
     at sofia.graphics.Polygon.get(Polygon.java:215)
     at sofia.graphics.internal.BayazitDecomposer.isReflex(BayazitDecomposer.java:391)
     at sofia.graphics.internal.BayazitDecomposer.decompose(BayazitDecomposer.java:91)
     at sofia.graphics.internal.BayazitDecomposer.decompose(BayazitDecomposer.java:230)
     at sofia.graphics.internal.BayazitDecomposer.<init>(BayazitDecomposer.java:53)
     at sofia.graphics.Polygon.convexDecomposition(Polygon.java:195)
     at org.awesomeco.trianglesmash.PolygonShape.createFixtures(PolygonShape.java:143)
     at sofia.graphics.Shape$1.run(Shape.java:100)
     at sofia.graphics.ShapeField.runOnceUnlocked(ShapeField.java:617)
     at sofia.graphics.Shape.createB2Body(Shape.java:96)
     at sofia.graphics.ShapeField.handleShapesAdded(ShapeField.java:571)
     at sofia.graphics.ShapeField.add(ShapeField.java:150)
     at sofia.graphics.ShapeView.add(ShapeView.java:399)
     at sofia.app.ShapeScreen.add(ShapeScreen.java:160)
     at org.awesomeco.trianglesmash.TriangleSmashScreen.initialize(TriangleSmashScreen.java:61)
     at java.lang.reflect.Method.invoke(Native Method)
     at java.lang.reflect.Method.invoke(Method.java:372)
     at sofia.internal.events.EventDispatcher$MethodTransformer.invoke(EventDispatcher.java:474)
     at sofia.internal.events.EventDispatcher.invokeTransformer(EventDispatcher.java:136)
     at sofia.internal.events.EventDispatcher.dispatch(EventDispatcher.java:109)
     at sofia.app.internal.ScreenMixin.invokeInitialize(ScreenMixin.java:561)
     at sofia.app.Screen$1.run(Screen.java:199)
     at android.os.Handler.handleCallback(Handler.java:739)
     at android.os.Handler.dispatchMessage(Handler.java:95)
     at android.os.Looper.loop(Looper.java:135)
     at android.app.ActivityThread.main(ActivityThread.java:5257)
     at java.lang.reflect.Method.invoke(Native Method)
     at java.lang.reflect.Method.invoke(Method.java:372)
     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:903)
     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698)
```

**Comment by allevato** [comment]
Unfortunately I don't think I'll have time in the foreseeable future to take a deep look into why you're getting that error; it's been a while since I've looked at that code, and it was ported from an example I found elsewhere on line (which describes the algorithm in more detail).

That being said, it's possible to work around it in this particular (triangle) case. For any three points there is a single unique triangle that passes through them, and a triangle can never be concave. That means we wouldn't actually need to run the decomposition in that situation. (It wouldn't surprise me if some part of the Bayazit algorithm, at least at the top level, assumes that it has more vertices due to that fact. Or it could just be buggy :) )

So a reasonable fix (which would be a bit more efficient too, in all likelihood) would be to simply not run the decomposition if the polygon has less than 4 vertices. In that case, the decomposition would simply be the polygon itself; the best place to do this would be to add a check in `Polygon.convexDecomposition` to return a singleton list containing a copy of `this`.
